### PR TITLE
Fix json parsing dependency introduced by mutijson config

### PR DESCRIPTION
**Pull request by xiaokangwang**
This PR removes the need to include json config parsing code in V2Fly.
The dependency for json parser is accidentally included when introducing mutijson file format.

### CODE REVIEW

**Review by vcptr — APPROVED** [review]

**Comment by forever8938** [comment]
Does this also fix the [issue](https://github.com/v2fly/v2ray-core/issues/247#issuecomment-706516982) that, no matter how many modules are commented out [here](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go), binary size remain unchanged as long as either of json or jsonem is included?

**Comment by xiaokangwang** [comment]
> Does this also fix the [issue](https://github.com/v2fly/v2ray-core/issues/247#issuecomment-706516982) that, no matter how many modules are commented out [here](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go), binary size remain unchanged as long as either of [json](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L62) or [jsonem](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L64) is included?

Yes, this also fix that issue based on my test.

**Comment by forever8938** [comment]
> > Does this also fix the [issue](https://github.com/v2fly/v2ray-core/issues/247#issuecomment-706516982) that, no matter how many modules are commented out [here](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go), binary size remain unchanged as long as either of [json](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L62) or [jsonem](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L64) is included?
> 
> Yes, this also fix that issue based on my test.

I just confirmed this issue persists when jsonem is used.

File: main/json/config_json.go [hunk 1]
```diff
@@ -4,11 +4,10 @@ package json
 
 import (
 	"io"
-
+	"os"
```
Comment (kslr) [inline main/json/config_json.go:7]: https://github.com/golangci/golangci-lint/issues/138

**Comment by xiaokangwang** [comment]
> > > Does this also fix the [issue](https://github.com/v2fly/v2ray-core/issues/247#issuecomment-706516982) that, no matter how many modules are commented out [here](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go), binary size remain unchanged as long as either of [json](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L62) or [jsonem](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L64) is included?
> > 
> > 
> > Yes, this also fix that issue based on my test.
> 
> I applied this PR but the issue persists when jsonem is used.

You should try to use json not jsonem. jsonem will introduce the dependency of all protocols.

**Comment by qjebbs** [comment]
> > > > Does this also fix the [issue](https://github.com/v2fly/v2ray-core/issues/247#issuecomment-706516982) that, no matter how many modules are commented out [here](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go), binary size remain unchanged as long as either of [json](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L62) or [jsonem](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L64) is included?
> > > 
> > > 
> > > Yes, this also fix that issue based on my test.
> > 
> > 
> > I applied this PR but the issue persists when jsonem is used.
> 
> You should try to use json not jsonem. jsonem will introduce the dependency of all protocols.

It comes to my mind that, If we use `map[string]interface{}` loader here(#348) ONLY, theoretically `jsonem` won't need to introduce all protocols, and not affect cropping. Is it right? But it would be a incompatible change.😅

**Comment by xiaokangwang** [comment]
> > > > > Does this also fix the [issue](https://github.com/v2fly/v2ray-core/issues/247#issuecomment-706516982) that, no matter how many modules are commented out [here](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go), binary size remain unchanged as long as either of [json](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L62) or [jsonem](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L64) is included?
> > > > 
> > > > 
> > > > Yes, this also fix that issue based on my test.
> > > 
> > > 
> > > I applied this PR but the issue persists when jsonem is used.
> > 
> > 
> > You should try to use json not jsonem. jsonem will introduce the dependency of all protocols.
> 
> It comes to my mind that, If we use `map[string]interface{}` loader here(#348) ONLY, theoretically `jsonem` won't need to introduce all protocols, and not affect cropping. Is it right? But it would be a incompatible change.sweat_smile

Yes, but internally json parser needs to import necessary package to get the definition of configure file.

**Review by kslr — APPROVED** [review]

**Comment by qjebbs** [comment]
> > > > > > Does this also fix the [issue](https://github.com/v2fly/v2ray-core/issues/247#issuecomment-706516982) that, no matter how many modules are commented out [here](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go), binary size remain unchanged as long as either of [json](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L62) or [jsonem](https://github.com/v2fly/v2ray-core/blob/master/main/distro/all/all.go#L64) is included?
> > > > > 
> > > > > 
> > > > > Yes, this also fix that issue based on my test.
> > > > 
> > > > 
> > > > I applied this PR but the issue persists when jsonem is used.
> > > 
> > > 
> > > You should try to use json not jsonem. jsonem will introduce the dependency of all protocols.
> > 
> > 
> > It comes to my mind that, If we use `map[string]interface{}` loader here(#348) ONLY, theoretically `jsonem` won't need to introduce all protocols, and not affect cropping. Is it right? But it would be a incompatible change.sweat_smile
> 
> Yes, but internally json parser needs to import necessary package to get the definition of configure file.

Forget about it.

You're right, it's only "THEORETICALLY". To accomplish, we need to do so much work, like:

- Load json into map, no into struct
- If any protocols registered, they go to the map, retrieve & parse their own part of data.

Take `Config.ReverseConfig` as example:

1. configMap keeps whole config data
1. ReverseConfig loader gets data from `m,ok:= configMap["reverse"]`
1. marshal from `m` to json `string`
1. unmarshal from string to struct `ReverseConfig`

> If ReverseConfig not registered, there's just no one cares about the data `configMap["reverse"]`

That requires changing of almost everything.

**Comment by kslr** [comment]
We can do these tasks in v5

**Pull request merged**